### PR TITLE
Add voter for workflow right

### PR DESCRIPTION
**Pull request by itkg-nanne**

### CODE REVIEW

File: WorkflowFunctionAdminBundle/Security/Authorization/Voter/WorkflowRightVoter.php [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    /**
+     * Checks if the voter supports the given class.
+     *
+     * @param string $class A class name
+     *
+     * @return bool true if this Voter can process the class
+     */
+    public function supportsClass($class)
+    {
+        return $class instanceof ContentInterface || $class instanceof NodeInterface;
```
Comment (nicolasThal) [inline WorkflowFunctionAdminBundle/Security/Authorization/Voter/WorkflowRightVoter.php]: you should create an unique interface for both of these entity

File: WorkflowFunctionAdminBundle/Tests/Security/Authorization/Voter/WorkflowRightVoterTest.php [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        $this->assertTrue($this->voter->supportsClass($class));
+    }
+
+    /**
+     * @return array
+     */
+    public function provideClassName()
+    {
+        return array(
+            array(Phake::mock('OpenOrchestra\ModelInterface\Model\ContentInterface')),
```
Comment (nicolasThal) [inline WorkflowFunctionAdminBundle/Tests/Security/Authorization/Voter/WorkflowRightVoterTest.php]: you could add some interfaces which should not be supported

Comment (itkg-nanne) [inline WorkflowFunctionAdminBundle/Tests/Security/Authorization/Voter/WorkflowRightVoterTest.php]: this test takes place in voter with stdClass and ::ABSTAIN.

File: WorkflowFunctionAdminBundle/Security/Authorization/Voter/WorkflowRightVoter.php [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+namespace OpenOrchestra\WorkflowFunctionAdminBundle\Security\Authorization\Voter;
+
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Authorization\Voter\VoterInterface;
+use OpenOrchestra\ModelInterface\Repository\ContentTypeRepositoryInterface;
+use OpenOrchestra\ModelInterface\Model\ContentInterface;
+use OpenOrchestra\WorkflowFunction\Repository\WorkflowRightRepositoryInterface;
+use OpenOrchestra\WorkflowFunction\Model\WorkflowRightInterface;
+use FOS\UserBundle\Model\UserInterface;
+
```
Comment (nicolasThal) [inline WorkflowFunctionAdminBundle/Security/Authorization/Voter/WorkflowRightVoter.php]: one empty line too much

File: WorkflowFunctionAdminBundle/Security/Authorization/Voter/WorkflowRightVoter.php [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+     *
+     * @return int either ACCESS_GRANTED, ACCESS_ABSTAIN, or ACCESS_DENIED
+     */
+    public function vote(TokenInterface $token, $object, array $attributes)
+    {
+        if (!$this->supportsClass($object)) {
+            return VoterInterface::ACCESS_ABSTAIN;
+        }
+        if (($user = $token->getUser()) instanceof UserInterface) {
+            $workflowRight = $this->workflowRightRepository->findOneByUserId($token->getUser()->getId());
```
Comment (nicolasThal) [inline WorkflowFunctionAdminBundle/Security/Authorization/Voter/WorkflowRightVoter.php]: you could directly use the $user variable and not $token->getUser()

File: WorkflowFunctionAdminBundle/Tests/Security/Authorization/Voter/WorkflowRightVoterTest.php [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        $this->assertEquals($isSupported, $this->voter->supportsClass($class));
+    }
+
+    /**
+     * @return array
+     */
+    public function provideClassName()
+    {
+        return array(
+            array(Phake::mock('stdClass'), false),
```
Comment (nicolasThal) [inline WorkflowFunctionAdminBundle/Tests/Security/Authorization/Voter/WorkflowRightVoterTest.php]: I would add also the class ContentTypeInterface or StatusableInterface which should not be supported

**Comment by nicolasThal** [comment]
:+1: with the little modification

**Comment by itkg-ngilain** [comment]
:+1:

File: BackofficeBundle/StrategyManager/AuthorizeStatusChangeManager.php [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    /**
+     * @param StatusableEvent $event
+     *
+     * @return bool
+     */
+    public function isGranted(StatusableEvent $statusableEvent)
+    {
+        $granted = true;
+        foreach ($this->strategies as $strategy) {
+            $granted = $granted && $strategy->isGranted($statusableEvent);
```
Comment (nicolasThal) [inline BackofficeBundle/StrategyManager/AuthorizeStatusChangeManager.php]: if the return of the strategy is false, you can exit directly with a return;

This way you will limit the work done

File: BackofficeBundle/StrategyManager/AuthorizeStatusChangeManager.php [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+        }
+        if ($granted) {
```
Comment (nicolasThal) [inline BackofficeBundle/StrategyManager/AuthorizeStatusChangeManager.php]: if there is an exit in the foreach, you don't need to put the check on the $granted variable

File: WorkflowFunctionAdminBundle/Security/Authorization/Voter/WorkflowRightVoter.php (hunk 4 above)
Comment (itkg-nanne) [inline WorkflowFunctionAdminBundle/Security/Authorization/Voter/WorkflowRightVoter.php]: How do you do that? Is it possible to drive first parameterr too?

File: WorkflowFunctionAdminBundle/Tests/Security/Authorization/Voter/WorkflowRightVoterTest.php (hunk 5 above)
Comment (itkg-nanne) [inline WorkflowFunctionAdminBundle/Tests/Security/Authorization/Voter/WorkflowRightVoterTest.php]: @nicolasThal 
How could you test for StatusableInterface as is_subclass_of(Phake::mock('OpenOrchestra\ModelInterface\Model\StatusableInterface'), 'OpenOrchestra\ModelInterface\Model\StatusableInterface') is true?

File: Backoffice/AuthorizeStatusChange/AuthorizeStatusChangeInterface.php [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+ * Interface AuthorizeStatusChangeInterface
+ */
+interface AuthorizeStatusChangeInterface
+{
+    /**
+     * @param StatusableEvent $event
+     *
+     * @return bool
+     */
+    public function isGranted(StatusableEvent $statusableEvent);
```
Comment (nicolasThal) [inline Backoffice/AuthorizeStatusChange/AuthorizeStatusChangeInterface.php]: The isGranted method should not accept a statusableEvent but a StatusableInterface and a StatusInterface in the signature

File: BackofficeBundle/EventSubscriber/UpdateStatusSubscriber.php [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    /**
+     * @param StatusableEvent $event
+     */
+    public function updateStatus(StatusableEvent $event)
+    {
+        $document = $event->getStatusableElement();
+        $toStatus = $event->getToStatus();
+        if ($this->authorizeStatusChangeManager->isGranted($event)) {
```
Comment (nicolasThal) [inline BackofficeBundle/EventSubscriber/UpdateStatusSubscriber.php]: so here you will change isGranted($event) to isGranted($document, $toStatus)

File: BackofficeBundle/StrategyManager/AuthorizeStatusChangeManager.php [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        $this->strategies[$strategy->getName()] = $strategy;
+    }
+
+    /**
+     * @param StatusableEvent $event
+     *
+     * @return bool
+     */
+    public function isGranted(StatusableEvent $statusableEvent)
```
Comment (nicolasThal) [inline BackofficeBundle/StrategyManager/AuthorizeStatusChangeManager.php]: same here

File: WorkflowFunctionAdminBundle/AuthorizeStatusChange/Strategies/WorkflowRightStrategy.php [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+        $this->roleRepository = $roleRepository;
+        $this->workflowFunctionRepository = $workflowFunctionRepository;
+    }
+
+    /**
+     * @param StatusableEvent $event
+     *
+     * @return bool
+     */
+    public function isGranted(StatusableEvent $event)
```
Comment (nicolasThal) [inline WorkflowFunctionAdminBundle/AuthorizeStatusChange/Strategies/WorkflowRightStrategy.php]: do not give the event, but all the argument in the event

**Comment by itkg-ngilain** [comment]
:+1:

**Comment by nicolasThal** [comment]
:+1:

**Pull request merged**